core/common/src/main/kotlin/app/k9mail/core/common/net/ssl/OkHttpTrustedCertificatesInstaller.kt

```diff
@@ -0,0 +1,18 @@
+package app.k9mail.core.common.net.ssl
+
+import okhttp3.OkHttpClient
+import okhttp3.tls.HandshakeCertificates
+
+fun OkHttpClient.Builder.installTrustedCertificates(
+    trustedCertificateProvider: TrustedCertificateProvider,
+): OkHttpClient.Builder {
+    val handshakeCertificates = HandshakeCertificates.Builder().apply {
+        trustedCertificateProvider.getCertificates().forEach { addTrustedCertificate(it) }
+
+        addPlatformTrustedCertificates()
+    }.build()
+
+    sslSocketFactory(handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager)
```

Comment:
> This is the part that I think has the potential to create problems. Therefore I'd like us to limit calling `OkHttpClient.Builder.sslSocketFactory()` to old Android versions.
> 
> Alternatively, the version check could be done at the call site to avoid calling `installTrustedCertificates()` for newer Android versions. That would probably also avoid some unnecessary class loading.

It there some known evidendence that using the `OkHttpClient.Builder.sslSocketFactory()` is problematic? If so it would be good to point to that issue or if not known to them, open a new one so it could be fixed.

Reply:
No. It's strictly about minimizing the attack surface. If we don't need to use additional code in a security critical path, we shouldn't.

Reply:
> If that's a concern, then I'm a little worried about our existing SSL implementation. We fiddle with the KeyStore, TrustManager and SSLContext quite a bit. And from a first glance, it looks dated. I would suggest focus the energy on them instead.

That's a valid concern. Especially since we don't have any tests for `DefaultTrustedSocketFactory`. That code has the upside of having been part of the security audit, though.
However, I don't see how any of that is relevant to this change. We want to do better than we did in the past.

> Also just using the Google recommended way of adding the certificates as network security config is basically doing the same. Just doesn't work prior API 24 and will add the certificate for every API level, even if it's already in the system defaults. So no extra API level check as suggested here.

The same principle applies. And since the network security config is loaded via the resource system, we could limit it to just affected Android versions.

Reply:
The same is true for the OkHttp implementation, they even have an active bug bounty programm. So I would trust their implementation more than ours.

But anyway, we need to decide on a strategy how to work with security related issues and against whom to protect. This would allow us to draw a line in terms of effort and time investment.

I would be more than happy to raise the bar and apply more security related aspect to everything we develop.

As you mentioned the security audit, I think there are still a couple of open proposals on how to improve in terms of security and I didn't see too much activity to address these. I'm happy that there is a change now.

Reply:
Also for complication, the `ISRG Root X2` certificate is only available from Android 14 on.

Reply:
I don't think that's a problem right now. It's cross-signed by X1 and by default X2 is included in the chain.

From <https://letsencrypt.org/certificates/>:
> All certificates signed by the ECDSA intermediate “E1” will come with a chain including an intermediate certificate whose Subject is “ISRG Root X2” and whose Issuer is “ISRG Root X1”. Almost all server operators will choose to serve this chain as it offers the most compatibility until ISRG Root X2 is widely trusted.

Reply:
Some more details for Android 14: [ISRG Root X2](https://issuetracker.google.com/issues/184856859)

Reply:
There's a cross sign from ISRG Root X2 to ISRG Root X1, but the cross signs are optional. Any server operator can choose to no serve them.

Reply:
Let's not expand the scope. The goal of this change is to keep Let's Encrypt certificates working on Android 7 and below. Not to increase Let's Encrypt compatibility in general.

Reply:
I think having better Let's Encrypt compatibility as a side effect of a patch should be highly welcome, as it's one of the largest certificate providers.

But I spend enough time on this topic. So I leave this as is and if you have a better idea, feel free to either create a pull-request or change this one.